laurent22/joplin
# Desktop: Improve beta editor support for the Rich Markdown plugin

# Summary

This pull request builds on the changes to the CodeMirror 5 emulation in https://github.com/laurent22/joplin/pull/9928 to make more features of the Rich Markdown plugin work in the CodeMirror 6 beta editor.

> [!NOTE]
>
> This pull request currently includes **all** changes from https://github.com/laurent22/joplin/pull/9928.
>
> It also requires this [Rich Markdown pull request](https://github.com/CalebJohn/joplin-rich-markdown/pull/58) to be merged.
>

Specifically, it,
1. Adds the `CodeMirror` class to the CodeMirror 6 editor. This allows existing Rich Markdown stylesheets to be applied to the editor.
   - Because CodeMirror 5 stylesheets are applied to `.CodeMirror` elements, this pull request adjusts several existing styles.
2. Adds support for the `styleActiveLine` CodeMirror 5 option
3. Support `editor.addWidget`
4. Support `editor.markText` (originally from https://github.com/laurent22/joplin/pull/9928)
5. Fixes `defineExtension` doesn't override existing extensions with the same name
6. Adjust how `coordsChar` handles the `page` mode
7. Adds the `CodeMirror-line` class to visible lines in the editor
8. Fixes an issue with loading CodeMirror 6 content scripts that contain inline comments

This pull request still requires changes to the Rich Markdown plugin. A pull request will be opened shortly.

# Testing

Regression testing (desktop):
1. Open the welcome notes in the beta editor
2. Verify that math is syntax highlighted
3. Verify that tables have a monospaced font
4. Verify that headers have a larger font size than non-headers
5. Verify that changing the default editor font in settings changes the editor font
6. Verify that changing the editor font size in settings changes the editor's font size

This has been successfully tested on Ubuntu 23.10.

Regression testing (mobile):
1. Open the welcome notes
2. Verify that math has syntax highlighting
3. Verify that tables have a monospaced font
4. Verify that headers have a larger size than nonheaders
5. Verify that the editor's font size can be changed in settings

This has been tested successfully on Android 7.
